storage_file/models/storage_file.py

```diff
@@ -82,49 +77,59 @@ def write(self, vals):
+            record.human_file_size = human_size(self.file_size)
+
+    def _build_relative_path(self, checksum):
+        self.ensure_one()
+        if not self.backend_id.filename_strategy:
+            raise UserError(_(
+                "The filename strategy is empty for the backend %s.\n"
+                "Please configure it") % self.backend_id.name)
+        if self.backend_id.filename_strategy == 'hash':
+            return checksum[:2] + '/' + checksum
+        elif self.backend_id.filename_strategy == 'name_with_id':
+            return "%s-%s%s" % (self.filename, self.id, self.extension)
 
-    def _prepare_meta_for_file(self, datas, private_path):
+    def _prepare_meta_for_file(self):
+        bin_data = base64.b64decode(self.datas)
+        checksum = hashlib.sha1(bin_data).hexdigest()
+        relative_path = self._build_relative_path(checksum)
         return {
-            'url': self.backend_id.sudo().get_public_url(private_path),
-            'checksum': hashlib.sha1(datas).hexdigest(),
-            'file_size': len(datas),
-            'private_path': private_path
+            'checksum': checksum,
+            'file_size': len(bin_data),
+            'relative_path': relative_path
             }
 
     @api.multi
     def _inverse_datas(self):
         for record in self:
-            b_decoded = base64.b64decode(record.datas)
-            private_path = self.backend_id.sudo().store(
-                record.name,
-                b_decoded,
-                is_public=True,
-                is_base64=False)
-            vals = record._prepare_meta_for_file(b_decoded, private_path)
-            record.write(vals)
+            record.write(record._prepare_meta_for_file())
+            record.backend_id.sudo().store(record.relative_path, record.datas)
 
     @api.multi
     def _compute_datas(self):
         for rec in self:
-            if not rec.url:
+            if self._context.get('bin_size'):
+                rec.datas = rec.file_size
+            elif rec.relative_path:
+                rec.datas = rec.backend_id.sudo().retrieve_data(
+                    rec.relative_path)
+            else:
                 rec.datas = None
+
+    @api.depends(
+        'backend_id.served_by', 'backend_id.base_url', 'relative_path')
+    def _compute_url(self):
+        for record in self:
+            if record.backend_id.served_by == 'odoo':
+                base_url = self.env['ir.config_parameter'].sudo()\
+                    .get_param('web.base.url')
+                record.url = \
+                    base_url + '/web/content/storage.file/%s/datas' % self.id
             else:
-                try:
-                    rec.datas = rec.backend_id.sudo().retrieve_data(
-                        rec.private_path)
-                except:
-                    _logger.error('Image %s not found', rec.url)
-                    rec.datas = None
-                    raise
+                record.url = "%s/%s" % (
```

Comment:
the construction of the url should be delegated to the backend.
You can't assume here any pattern.